design/src/rfcs/rfc0022_refine_builder.md

```diff
@@ -0,0 +1,556 @@
+The issue won't be surfaced by a `cargo check` or `cargo build` invocation, as it happens with the typestate builder approach.  
+It should be surfaced by executing the application test suite, assuming that the developer has written at least a single integration test - e.g. a test that passes a request to the `call` method exposed by `PokemonService` or launches a full-blown instance of the application which is then probed via an HTTP client.  
+
+If there are no integration tests, the issue won't be detected on the developer machine nor in CI.  
+Nonetheless, it is unlikely to cause any end-user impact even if it manages to escape detection and reach production. The deployment will never complete if they are using a progressive rollout strategy: instances of the new version will crash as soon as they are launched, never getting a chance to mark themselves as healthy; all traffic will keep being handled by the old version, with no visible impact on end users of the application.
+
+Given the above, we think that the impact of a runtime error is low enough to be worth exploring designs that do not guarantee compile-safety for the builder API[^further-dev-productivity-improvements].  
+
+### Providing clear feedback
+
+Moving from a compile-time error to a runtime error does not require extensive refactoring.  
+The definition of `PokemonServiceBuilder` goes from:
+
+```rust
+pub struct PokemonServiceBuilder<
+    Op1,
+    Op2,
+    Op3,
+    Op4,
+    Op5,
+    Op6,
+    Exts1 = (),
+    Exts2 = (),
+    Exts3 = (),
+    Exts4 = (),
+    Exts5 = (),
+    Exts6 = (),
+    Pl = aws_smithy_http_server::plugin::IdentityPlugin,
+> {
+    check_health: Op1,
+    do_nothing: Op2,
+    get_pokemon_species: Op3,
+    get_server_statistics: Op4,
+    capture_pokemon: Op5,
+    get_storage: Op6,
+    #[allow(unused_parens)]
+    _exts: std::marker::PhantomData<(Exts1, Exts2, Exts3, Exts4, Exts5, Exts6)>,
+    plugin: Pl,
+}
+```
+
+to:
+
+```rust
+pub struct PokemonServiceBuilder<
+    Op1,
+    Op2,
+    Op3,
+    Op4,
+    Op5,
+    Op6,
+    Exts1 = (),
+    Exts2 = (),
+    Exts3 = (),
+    Exts4 = (),
+    Exts5 = (),
+    Exts6 = (),
+    Pl = aws_smithy_http_server::plugin::IdentityPlugin,
+> {
+    check_health: Option<Op1>,
+    do_nothing: Option<Op2>,
+    get_pokemon_species: Option<Op3>,
+    get_server_statistics: Option<Op4>,
+    capture_pokemon: Option<Op5>,
+    get_storage: Option<Op6>,
+    #[allow(unused_parens)]
+    _exts: std::marker::PhantomData<(Exts1, Exts2, Exts3, Exts4, Exts5, Exts6)>,
+    plugin: Pl,
+}
+```
+
+All operation fields are now `Option`-wrapped.  
+We introduce a new `MissingOperationsError` error to hold the names of the missing operations and their respective setter methods:
+
+```rust
+#[derive(Debug)]
+pub struct MissingOperationsError {
+    service_name: &'static str,
+    operation_names2setter_methods: HashMap<&'static str, &'static str>,
```

Comment:
We can probably just use a `TinyMap` or even `SmallVec/TinyVec` here, right?

Reply:
Possibly! I kept the code simple in the RFC, but we can of course refine the internal implementation details when actually moving forward with the implementation.